fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/explain/DotExporter.java

```diff
@@ -303,48 +346,67 @@ protected void renderFooter(final ExporterContext context) {
+    }
 
-        if (htmlLabel != null) {
-            out.print("label=");
-            out.print(htmlLabel);
-            out.print(" ");
+    @SuppressWarnings({"squid:S3358", "unchecked"})
+    public String getNodeLabel(@Nonnull final Attribute name, final Attribute details, final Map<String, Attribute> nodeAttributes) {
+        if (details == null || ((List<?>)details.getReference()).isEmpty()) {
+            return "<<table border=\"0\" cellborder=\"1\" cellspacing=\"0\" cellpadding=\"8\"><tr><td>" +
+                   escaper.escape(name.getReference().toString()) +
+                   "</td></tr></table>>";
         }
 
-        for (Map.Entry<String, Attribute> entry : attributes.entrySet()) {
-            final String key = entry.getKey();
-            if (doNotRenderVerbatim.contains(key)) {
-                // already handled by special case above
-                continue;
-            }
-            renderAttribute(context, key, entry.getValue(), " ");
-        }
-        out.print("]");
+        final String detailsString =
+                ((List<?>)details.getReference())
+                        .stream()
+                        .map(Object::toString)
+                        .map(escaper::escape)
+                        .map(detail ->
+                                substituteVariables(detail,
+                                        nodeAttributes,
+                                        attribute -> attribute == null
+                                                     ? "<b>undefined</b>"
+                                                     : (attribute.getReference() instanceof Collection)
+                                                       ? escapeCollection((Collection<Attribute>)attribute.getReference())
+                                                       : escaper.escape(attribute.getReference().toString())))
+                        .map(detail -> "<tr><td>" + detail + "</td></tr>")
+                        .collect(Collectors.joining());
+
+        return "<<table border=\"0\" cellborder=\"1\" cellspacing=\"0\" cellpadding=\"8\"><tr><td>" +
+               escaper.escape(name.getReference().toString()) + "</td></tr>" +
+               detailsString + "</table>>";
+    }
+
+    private String escapeCollection(final Collection<Attribute> attributes) {
+        return "[" + attributes.stream().map(a -> escaper.escape(a.toString())).collect(Collectors.joining(", ")) + "]";
     }
 
-    public String getLabel(@Nonnull final Attribute name, final Attribute expression) {
-        if (expression == null || expression.toString().isEmpty()) {
-            return "<<table border=\"0\" cellborder=\"1\" cellspacing=\"0\" cellpadding=\"8\"><tr><td>" + name + "</td></tr></table>>";
+    private String substituteVariables(final String detail,
+                                       final Map<String, Attribute> nodeAttributes,
+                                       final Function<Attribute, String> toStringFn) {
```

Comment:
```suggestion
    private String substituteVariables(@Nonnull final String detail,
                                       @Nonnull final Map<String, Attribute> nodeAttributes,
                                       @Nonnull final Function<Attribute, String> toStringFn) {
```

Reply:
done for every method in that hierarchy.